### ISSUE TITLE
Checking for musl results in java.nio.file.FileSystemException: /proc/self/map_files: Operation not permitted

### ISSUE THREAD

**Issue by noahlz** [xerial/sqlite-jdbc#706]
Reference #703 and #704

My (admittedly old, please forgive me) version of Ubuntu Linux (Ubuntu 14.04 LTS) cannot use the sqlite-jdbc we built for Archlinux from the above PR branch.

/proc/self/map_files is not readable by non-root (sudo) users

```
scala> import java.nio.file.Paths
import java.nio.file.Paths

scala> import java.nio.file.Files
import java.nio.file.Files

scala> val path = Paths.get("/proc/self/map_files")
path: java.nio.file.Path = /proc/self/map_files

scala> val stream = Files.list(path)
stream: java.util.stream.Stream[java.nio.file.Path] = java.util.stream.ReferencePipeline$Head@7d37d083

scala> stream.map(_.toRealPath()).forEach(println)
java.io.UncheckedIOException: java.nio.file.FileSystemException: /proc/self/map_files: Operation not permitted
  at java.nio.file.Files$2.hasNext(Files.java:3462)
  at java.util.Iterator.forEachRemaining(Iterator.java:115)
  at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
  at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
  at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
  at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
  at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
  at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
  ... 31 elided
Caused by: java.nio.file.FileSystemException: /proc/self/map_files: Operation not permitted
  at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
  at sun.nio.fs.UnixException.asIOException(UnixException.java:111)
  at sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.readNextEntry(UnixDirectoryStream.java:171)
  at sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.hasNext(UnixDirectoryStream.java:201)
  at java.nio.file.Files$2.hasNext(Files.java:3460)
  ... 39 more
```

I confirmed this by attempting to read the /proc/self/map_files with and without sudo:

```
$ cd /proc/self
$ ls map_files
ls: reading directory map_files: Operation not permitted
$ sudo ls map_files
400000-4f0000              7ff088fda000-7ff088fdb000  7ff0893ff000-7ff089400000  7ff089cd1000-7ff089cd2000
6ef000-6f0000              7ff088fdb000-7ff088fdc000  7ff089400000-7ff089705000  7ff089cd2000-7ff089cf7000
6f0000-6f9000              7ff088fdc000-7ff088ff3000  7ff089705000-7ff0898c3000  7ff089cf7000-7ff089ef6000
7ff088bc5000-7ff088bcf000  7ff088ff3000-7ff0891f2000  7ff0898c3000-7ff089ac3000  7ff089ef6000-7ff089efa000
7ff088bcf000-7ff088dce000  7ff0891f2000-7ff0891f3000  7ff089ac3000-7ff089ac7000  7ff089efa000-7ff089efb000
7ff088dce000-7ff088dcf000  7ff0891f3000-7ff0891f4000  7ff089ac7000-7ff089ac9000  7ff089efb000-7ff089f1e000
7ff088dcf000-7ff088dd0000  7ff0891f6000-7ff0891ff000  7ff089ace000-7ff089ad1000  7ff08a116000-7ff08a11d000
7ff088dd0000-7ff088ddb000  7ff0891ff000-7ff0893fe000  7ff089ad1000-7ff089cd0000  7ff08a11d000-7ff08a11e000
7ff088ddb000-7ff088fda000  7ff0893fe000-7ff0893ff000  7ff089cd0000-7ff089cd1000  7ff08a11e000-7ff08a11f000
```

"Upgrade your distribution / use a Mac" is a perfectly valid response here. At the worst, hoping to assist someone who encounters a similar issue with this release.

**Comment by noahlz** [comment]
Original stack trace coming from the library:

```
Caused by: java.io.UncheckedIOException: java.nio.file.FileSystemException: /proc/self/map_files: Operation not permitted
        at java.nio.file.Files$2.hasNext(Files.java:3462)
        at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449)
        at org.sqlite.util.OSInfo.isMusl(OSInfo.java:127)
        at org.sqlite.util.OSInfo.translateOSNameToFolderName(OSInfo.java:226)
        at org.sqlite.util.OSInfo.getOSName(OSInfo.java:108)
        at org.sqlite.util.OSInfo.getNativeLibFolderPathForCurrentOS(OSInfo.java:104)
        at org.sqlite.SQLiteJDBCLoader.loadSQLiteNativeLibrary(SQLiteJDBCLoader.java:348)
        at org.sqlite.SQLiteJDBCLoader.initialize(SQLiteJDBCLoader.java:65)
        at org.sqlite.core.NativeDB.load(NativeDB.java:56)
        at org.sqlite.SQLiteConnection.open(SQLiteConnection.java:254)
        ... 59 more
Caused by: java.nio.file.FileSystemException: /proc/self/map_files: Operation not permitted
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
        at sun.nio.fs.UnixException.asIOException(UnixException.java:111)
        at sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.readNextEntry(UnixDirectoryStream.java:171)
        at sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.hasNext(UnixDirectoryStream.java:201)
        at java.nio.file.Files$2.hasNext(Files.java:3460)
        ... 76 more
```

**Comment by gotson** [comment]
I am no Linux expert, but we should try to understand what's happening from the OS point of view before doing anything else.

Is that something you can enquire about on one of the stack exchange websites maybe?

**Comment by noahlz** [comment]
This seems relevant:

"Why do files in /proc/self end up being owned by root if a program has its setuid bit set?"
https://stackoverflow.com/q/46897570/7507

And this

"What causes /proc/<pid>/* resources to become owned by root, despite the procs being launched as a normal user?"
https://unix.stackexchange.com/q/133922/64902

It seems that the native code (the Java binary in this case, I suppose), needs to be compiled with a "dumpable" flag. It seems reasonable that it would not have this flag enabled.

Perhaps there another way to resolve the platform architecture?

**Comment by gotson** [comment]
That's insightful, but it doesn't explain why it only happens on your machine though.

Could it be something with your setup?

**Comment by noahlz** [comment]
Having googled a bit on this, it seems to me that there was a change in the sqlite binary some time in the past few years, namely to disable the PR_SET_DUMPABLE flag...

I have a work-around for now, so this is not a blocker. Perhaps someone else will have a similar issue and come across this post.

**Comment by gotson** [comment]
What I'm wondering is why it only happens to you while many people run this on various Linux flavors without problem.

Is that something specific to the kernel version used in Ubuntu 14.04?

**Comment by pyckle** [comment]
Reproduced in CentOS 7. Fix committed in PR for 3.37.0 https://github.com/xerial/sqlite-jdbc/pull/710

**Comment by gotson** [comment]
> Reproduced in CentOS 7. Fix committed in PR for 3.37.0 #710

This doesn't really fix the problem, it only ignores the error for older Ubuntu and centos. It's good not to crash, but it doesn't solve the underlying problem. 

If the problem happens also on older alpine, then it will always resolve to not being musl and not work.

To understand what happens we would need to understand whether this is coming from the JDK bundled with those distros or not, and whether it's a kernel or distro behavior.

**Comment by pyckle** [comment]
Hi @gotson,

Thanks for the reply. It seems that map_files was added in linux 3.3. In newer kernels, permissions to use it require ptrace read. See: https://man7.org/linux/man-pages/man5/proc.5.html

>               Permission to access this file is governed by a ptrace
>               access mode PTRACE_MODE_READ_FSCREDS check; see ptrace(2).

However, the CentOS7 proc man page make no mention of this permission, and deny ordinary users access to this folder. I think it is reasonable to assume that other older kernels function similarly.

I don't have a better fix for this issue. Maybe falling back to the previous check for alpine linux is better than just assuming not musl is better behavior?

With that said, falling back in the case of exception to assuming not musl is better behavior than failing to run on older kernels. Is running sqlite-jdbc on top of an old linux kernel with musl a common use case, or even a use case at all? If so, perhaps a system property ("org.xerial.sqlite.forceMusl") is an acceptable for this rare case?

Lastly, the way this code is written, this function will run on AIX, and if it throws an exception, AIX would also be broken. This perhaps should be fixed too.

**Comment by gotson** [comment]
> Thanks for the reply. It seems that map_files was added in linux 3.3. In newer kernels, permissions to use it require ptrace read. See: https://man7.org/linux/man-pages/man5/proc.5.html
> 
> >

Thanks for the investigation. If that is correct, then:
- running as non-root on any Ubuntu with a kernel 3.3 or higher should fail. However this is not the case. The Github Actions tests run as non-root, and don't fail.
- running as non-root on any Alpine with a kernel 3.3 or higher should fail. However this is not the case. I tested this again on my machine, and this works.

This may be however because we are using Docker and not bare-metal or VMs, that may have some impact on the Kernel.



> I don't have a better fix for this issue. Maybe falling back to the previous check for alpine linux is better than just assuming not musl is better behavior?

I think that's a reasonable course of action, if not perfect, at least it will avoid crashes.

> If so, perhaps a system property ("org.xerial.sqlite.forceMusl") is an acceptable for this rare case?

I was thinking of something along those lines too, some kind of manual override. We can always add it later if someone complains.

> Lastly, the way this code is written, this function will run on AIX, and if it throws an exception, AIX would also be broken. This perhaps should be fixed too.

good call!

**Comment by pyckle** [comment]
Hi,

I think my message wasn't clear regarding the map_files interface. Let me try again (:

The interface itself was added to /proc in Linux 3.3. That means any Linux kernel before Linux 3.3 will not have /proc/<pid>/map_files for any user, root or non-root. In these kernels, this check will always fail, regardless of the user's permissions.

The ptrace permissions which allows non-root users to access this interface was added/changed sometimes later. The manpage for proc in CentOS 7 (RHEL7) does not mention anything about ptrace permissions in the map_files section. It simply denies access to map_files to non-root users. This is in Linux 3.10, but with Red Hat's patches (I haven't investigated whether they changed this). The permissions must have been added later.

Even in later kernels, a user can be denied even ptrace read access, and this would cause this code to fail. I suspect that this is an unusual configuration, but likely exists in some security hardened environments.

I don't think running in a docker container affects these user permissions vs. bare metal/vm. Docker just filters parts of /proc out that aren't relevant to the container.

In any case, we could figure out exactly where the procfs map_files permissions changed in the Linux source (it's a git blame away), but I don't see a point in doing so. Even kernels before 3.3 should be supported (otherwise building the x86/x86_64 images on top of a CentOS5 userspace would be silly).

I'm happy to add a system property into this PR for forcing musl on or off - just let me know what the conventions for property names are.

Cheers, and thanks for all of your feature contributions to SQLite - they come in handy (:

**Comment by gotson** [comment]
Thanks a lot for that very exhaustive explanation, I understand better now!

Xerial is not available at the moment, we can probably sit this out until he comes back. I'm nto sure we need the system property right now, and apart from the one that exists for the temp file location or native lib location, I don't think there are any yet.

**Comment by pyckle** [comment]
Perfect - sounds good. Let me know if there's anything else I can to do to get these improvements merged.

**Comment by gotson** [comment]
> Perfect - sounds good. Let me know if there's anything else I can to do to get these improvements merged.

I'm just commenting here and there but I don't have any rights on the repo, only xerial can merge those and he is still on parental leave from what I understand, so we'll just have to wait.
